src/sentry/static/sentry/app/components/hovercard.jsx

```diff
@@ -70,6 +79,71 @@ class Hovercard extends React.Component {
     return currentClasses.length > 0 ? currentClasses : classes;
   }
 
+  // figure out if we're dealing with hovercard-left or hovercard-right (in which case we don't
+  // want to center the hovercard) or hovercard or hovercard-bottom (in which case we do)
+  getMargin(classes) {
+    if (!this.cardElement || !this.state.visible) return {};
+
+    const rect = this.cardElement.getBoundingClientRect();
+
+    const rectWidth = this.isMultipleLines() ? rect.width : this.getSingleLineWidth();
+
+    if (classes.indexOf('left') !== -1 || classes.indexOf('right')) {
+      return -rectWidth / 2;
+    } else {
+      return 0;
+    }
+  }
+
+  isMultipleLines() {
+    if (!this.cardElement || !this.state.visible) return {};
+
+    // spans will create a new bounding box for each line of text
+    // if the text wraps, so this is a way to test for text wrapping
+    return this.cardBodySpan.getClientRects().length > 1;
+  }
+
+  //computed styles come back like "10px", so convert to the int 10
+  stylePxToNum(px) {
+    return parseInt(px.slice(0, -2), 10);
+  }
+
+  //this will be the width of the overall hovercard, based on the
+  //contents of the span in the body and all padding/borders around it
+  getSingleLineWidth() {
```

Comment:
you should be able to lean on the dom to do most of this work for you. just get the total width of whatever the outermost element is (probably `cardElement`?)

```js
const totalWidth = this.cardElement.clientWidth;
```

reading styles is kind of a hopeless battle—for instance adding a max-width, min-width, margin, etc. would break this.

Reply:
Have been banging my head against versions of that for 10 days now, on and off. I agree that the style-checking is brittle and real real ugly. I feel like it ought to be possible to get the browser to figure this out for me but I feel like I've tried every possible variation to no avail... :-(